### ISSUE TITLE
CleanMyMac over-polling our AppCast

### ISSUE THREAD

**Issue by psulak** [sparkle-project/Sparkle#2071]
Hi All -

We host our Sparkle AppCast on our own Server. Recently, our logs have shown a CRAZY amount of polling coming from IP's with User-Agents purporting to be CleanMyMac. I'm talking 10's of thousands of hits from a single IP in just ONE DAY. And this is happening from multiple IP's. After our AppCast is read, our Release Notes file is also ready by the same address. This has led us to believe that the issue is caused by CleanMyMac, and is not something like a DDOS attack disguising itself as CMM.

I've tried reaching out to the folks at CleanMyMac. But so far I haven't been able to talk to anyone on their engineering team. Is anyone here in contact with the CleanMyMac people about this? I'd be very surprised if ours was the only AppCast affected by this. Thanks!

A similar issue got mentioned in issue 1986 here: 
https://github.com/sparkle-project/Sparkle/issues/1986

Quick update: I've got CleanMyMac installed here. I went ahead and installed Little Snitch and I'm definitely seeing a lot of traffic between CMM and our Server. Along with CMM and all the other Sparkle-based apps on my Mac!

Update 2: It seems like I can actually TRIGGER CleanMyMac to do this on a whim. All I have to do is search my /Applications folder for specific files. I have no idea why a file search would cause CleanMyMac to do this. This is truly bizarre...

**Comment by zorgiepoo** [comment]
Several Mac developers have filed reports to them including herf ([via twitter](https://twitter.com/herf/status/1458920297650798602) from #1986). The CleanMyMac folk were supposed to fix a while back in an update, but if you can still repro it, maybe they did not really fix it. And users can still potentially be running an old version of CleanMyMac.

If you can post repro steps here (and/or to them) with their latest version, I can try contacting an engineer I know that is in close contact with them and reference them here. (The other annoying thing is that they may not be setting a custom user agent string to note that the update checks are coming from them..)

We also have [documentation](https://sparkle-project.org/documentation/publishing/#publishing-an-update) on setting the feed URL programmatically via a delegate method, instead of exposing the feed in the app bundle, if you want to go that route. This only adds a little bit of obscurity because 3rd party updaters like these can do some manual labor to find your feed URL and surely they likely do custom decision handling on app to app basis.

**Comment by stefanfuerst** [comment]
Confirmed. They seem to be sniffing your users network traffic. I have dynamic parameters generated at runtime and they are used by CMM

**Comment by storyer** [comment]
Hi Phil, that's Volodymyr from CleanMyMac here; please apologize for the issue that happened.

Thanks a lot for reporting the case. Yes, we've located its origin and are working on the fix right now. We'll do our best to release it in the nearest future.

Let me keep you posted on that. 

Best,
Volodymyr

**Comment by stefanfuerst** [comment]
As a developer **don't** configure your URL in the Info.plist. 
Also, don't use setFeedURL: on the SUUpdate, as this will write your URL to your UserDefaults.
```
you should consider implementing
-[SPUUpdaterDelegate feedURLStringForUpdater:] instead of using this method.
```

**Comment by lrrrs** [comment]
I can confirm this.
Removing the feed URL from the `Info.plist`

```
SUUpdater.shared().feedURL = nil
SUUpdater.shared().delegate = self
```
and implementing the delegate
```
func feedURLString(for updater: SUUpdater) -> String?
```

Nothing is stored in the UserDefaults, CleanMyMac can't read the feed URL. 
Peace restored.

**Comment by psulak** [comment]
> Hi Phil, that's Volodymyr from CleanMyMac here; please apologize for the issue that happened.
> 
> Thanks a lot for reporting the case. Yes, we've located its origin and are working on the fix right now. We'll do our best to release it in the nearest future.
> 
> Let me keep you posted on that.
> 
> Best, Volodymyr

Hi Volodymyr!

That is wonderful news! I currently have a tech support incident logged with MacPaw. If you need any more information from me, please feel free to contact me there. Or I can post info here. Whichever is easiest:)

**Comment by psulak** [comment]
> Several Mac developers have filed reports to them including herf ([via twitter](https://twitter.com/herf/status/1458920297650798602) from #1986). The CleanMyMac folk were supposed to fix a while back in an update, but if you can still repro it, maybe they did not really fix it. And users can still potentially be running an old version of CleanMyMac.

Thankfully it looks like the CleanMyMac/MacPaw guys are now on it!

> I can confirm this. Removing the feed URL from the `Info.plist`

Yes! We figured this out as well. Unfortunately this only helps for new builds. But we've got lots of users who have old builds of our software on their Macs. They either can't update (due to being OS-locked) or don't update (due to not running our app). Either way, all those old builds with Sparkle AppCast URLs in their Info.plists get found by CMM.

**Comment by zorgiepoo** [comment]
> I can confirm this. Removing the feed URL from the `Info.plist`
> 
> ```
> SUUpdater.shared().feedURL = nil
> SUUpdater.shared().delegate = self
> ```
> 
> and implementing the delegate fixes this.

If you use Sparkle 1, make sure the delegate is set right away by the way like via an IBOutlet or in applicationDidFinishLaunching: (2.x enforces this. Essentially in the same runloop cycle the updater is instantiated in). If Sparkle asks for the feed, and nothing is found, then bam.

**Comment by lrrrs** [comment]
> Yes! We figured this out as well. Unfortunately this only helps for new builds. But we've got lots of users who have old builds of our software on their Macs. They either can't update (due to being OS-locked) or don't update (due to not running our app). Either way, all those old builds with Sparkle AppCast URLs in their Info.plists get found by CMM.

Same applies for users who don't update CMM. Maybe CMM has a feature flag, remote trigger, etc. that can disable the updater? 🤷‍♂️ @storyer

**Comment by storyer** [comment]
Hey Irrrs, thanks for the question; let me get back to it in a while.

We've already released an updated version with the fix of that. So once customers are updated, that should not happen again. Thanks for letting us know about the case, and sorry for any inconvenience caused.

**Comment by psulak** [comment]
> We've already released an updated version with the fix of that. So once customers are updated, that should not happen again.

Volodymyr, is the fix in CMM v4.10.1?

I updated to v4.10.1 about 2 hours ago. I've done some quick tests here, and it appears that I'm no longer able to trigger CMM to poll various AppCasts simply by searching my /Applications folder. That's great news!

**Comment by psulak** [comment]
I think I'm going to go ahead and close this, since MacPaw seems to have fixed the issue:)
